osxkeychain.go

```diff
@@ -108,32 +109,50 @@ func (ke keychainError) Error() string {
 	return fmt.Sprintf("keychainError with unknown error code %d", C.OSStatus(ke))
 }
 
-func AddGenericPassword(attributes *GenericPasswordAttributes) error {
-	if err := attributes.CheckValidity(); err != nil {
-		return err
+func AddGenericPassword(attributes *GenericPasswordAttributes) (err error) {
+	if err = attributes.CheckValidity(); err != nil {
+		return
 	}
 
-	serviceName := C.CString(attributes.ServiceName)
-	defer C.free(unsafe.Pointer(serviceName))
+	var serviceNameString C.CFStringRef
+	if serviceNameString, err = _UTF8StringToCFString(attributes.ServiceName); err != nil {
+		return
+	}
+	defer C.CFRelease(C.CFTypeRef(serviceNameString))
 
-	accountName := C.CString(attributes.AccountName)
-	defer C.free(unsafe.Pointer(accountName))
+	var accountNameString C.CFStringRef
+	if accountNameString, err = _UTF8StringToCFString(attributes.AccountName); err != nil {
+		return
+	}
+	defer C.CFRelease(C.CFTypeRef(accountNameString))
 
-	password := unsafe.Pointer(C.CString(attributes.Password))
-	defer C.free(password)
+	dataBytes := C.CFTypeRef(bytesToCFData(attributes.Password))
+	defer C.CFRelease(dataBytes)
 
-	errCode := C.SecKeychainAddGenericPassword(
-		nil, // default keychain
-		C.UInt32(len(attributes.ServiceName)),
-		serviceName,
-		C.UInt32(len(attributes.AccountName)),
-		accountName,
-		C.UInt32(len(attributes.Password)),
-		password,
-		nil,
-	)
+	query := map[C.CFTypeRef]C.CFTypeRef{
+		C.kSecClass:       C.kSecClassGenericPassword,
+		C.kSecAttrService: C.CFTypeRef(serviceNameString),
+		C.kSecAttrAccount: C.CFTypeRef(accountNameString),
+		C.kSecValueData:   dataBytes,
+	}
 
-	return newKeychainError(errCode)
+	access, err := createAccess(attributes.ServiceName, attributes.TrustedApplications)
+	if err != nil {
+		return
+	}
+
+	if access != nil {
+		defer C.CFRelease(C.CFTypeRef(access))
+		query[C.kSecAttrAccess] = C.CFTypeRef(access)
+	}
+
+	queryDict := mapToCFDictionary(query)
+	defer C.CFRelease(C.CFTypeRef(queryDict))
+
+	errCode := C.SecItemAdd(queryDict, nil)
+
+	err = newKeychainError(errCode)
```

Comment:
The type of the return value of FindGenericPassword should probably be consistent with GenericPasswordAttributes.Password